### ISSUE TITLE
XSLT 3

### ISSUE THREAD

**Issue by msb81** [ebeshero/DHClass-Hub#233]
Hey guys, I'm having some problems with XSLT 3. I have the whole first column of questions done i can't figure out how to get to the next column of yes and so on. 

Can anyone help?

**Comment by RJP43** [comment]
@msb81 can you post your code, please?

**Comment by RJP43** [comment]
@msb81 I mean without looking at your code if you have correctly grabbed the questions (by selecting the elements that have the attribute that indicates question) then for the table cells (`<td>`) that grab the responses will have a similar format except taking a step further to get the `@n` value. So like this:
`<td><xsl:apply-templates select="element[@attribute='response looking for']/@n"/></td>`

Remember to reference our [expanded class-example](https://github.com/ebeshero/DHClass-Hub/blob/master/Class-Examples/XSLT/XSL-to-Table_Example/XSL-to-HTML-TableOutput.xsl) if you get stuck on where things are supposed to be placed in regards to the one-to-one versus many-to-one setup.

**Comment by RJP43** [comment]
@msb81 I would be happy to help you further if you could either post your code or give me a more specific question on what you are confused about.

**Comment by ebeshero** [comment]
@msb81 I'm on campus in my office this afternoon until 5ish if you want to come by! I'm in FOB 204.

**Comment by etj27** [comment]
He is taking a while to respond so I will respond instead, since I feel like I am further behind than him and I think I'm just struggling on exactly what to do.  

`<
    <xsl:template match="/">
        <html>
            <head>
                <title>
                    <h1> Sanitary Conditions of Workshops and Factories in New York City</h1>
                </title>
            </head>
            <body>
                <h1>Sanitary Conditions of Workshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                       
                    </tr>
                   <xsl:apply-templates select="descendant::f[@name='question']"/>
                    
                </table>
                <h2>Table 2: Sources of Offensive Odors</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Water Closet</th>
                        <th>Other Sources</th>
                        <th>Total Responses</th>
                    </tr>
                    
                </table>
                
                <h3>Table 3: Standing or Sitting at Work</h3>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Sit</th>
                        <th>Stand</th>
                        <th>Optional</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                
                </table>
               
            </body>
        </html>
    </xsl:template>



</xsl:stylesheet>`

So this is my take at the code so far.  I'm not even sure if I got the basic structure right.  So one thing i noticed is, okay I got some questions and they aren't sorted according to the amount of tables there are supposed to be which is whatever, I'm just trying to get some stuff done before i have to go to work after class.  So from here I actually don't know where to go.  I have the questions, so I guess i need to get the responses and everything else.  I guess my main question is I know what code i actually need but I don't know exactly where to put it.  Do I need more code after the table or do i go straight into more templates?  I tried looking at the in class example but it didn't really give me anything to help since for me the format of what each xslt is coding is vastly different

I think one of the reasons i'm stuck on questions is because there is no division between tables in the xml document and i can't figure out how to separate them and put them into different tables.  

So I guess for me I just need to figure out the skeleton of the code to get to business.  Hopefully this post makes some sense I'm actually in another class atm

**Comment by RJP43** [comment]
@etj27 (@msb81 is Maddee... lol)

Although the XML for the class example and the assignment are drastically different the XSLT setup is similar. You have done well here giving the HTML tags that you know appear only once inside of the document node match. Your apply-templates for each table will need to grab only the information that pertains to each of those unique tables. So in your first table setup your `<xsl:apply-templates select="descendant::f[@name='question']"/>` is saying in this table you want to process all the descendant `<f>` elements that have an `@name='question'` attribute. Now you are on the right track with this, but you want to be more allowing yet also more specific. In the table where you are giving the data on yes and no questions you want to process the contents of the `<fs>` elements that have descendant `<f>` elements that specifically have the kinds of responses present in that table (so either Yes, No, or Yes_but_fined for the first table) So each table will have their own `<xsl:apply-templates>` selection that is specific to the kind of responses be represented by that table. In our [class-example](https://github.com/ebeshero/DHClass-Hub/blob/master/Class-Examples/XSLT/XSL-to-Table_Example/XSL-to-HTML-TableOutput.xsl) (which I added to significantly after class) notice I marked each template match (after the first document node match) to indicate what `<xsl:apply-templates>` selection from the document node it matches. Each table will get one `<xsl:apply-templates>` selection in the source node template match and then a corresponding template match where you give the contents of the table row and each table cell.  The `<xsl:apply-templates>` selection in the source document is where you say HEY this table is going to have this specific info processed in it and the corresponding `xsl:template @match` (that follows the document node template match)  gives the directions of how that info is processed.

**Comment by setriplette** [comment]
HALP! When I was working with this earlier with Dr. B it was outputting more numbers, but as I've been working on the `<td>` elements for the second and third tables I'm coming up with an ambiguous rule match error and the numbers I used to output are now not outputting. I can also tell that I'm taking the sum of the wrong thing to get a total of responses but I don't know what the right thing would be.
`<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xpath-default-namespace="http://www.tei-c.org/ns/1.0"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    exclude-result-prefixes="xs math"
    xmlns="http://www.w3.org/1999/xhtml"
    version="3.0">
    <xsl:output method="xhtml" encoding="utf-8" doctype-system="about:legacy-compat"
        omit-xml-declaration="yes"/>
    <xsl:template match="/">
        <html>
            <head>
                <title>Sanitary Conditions NYC</title>
            </head>
            <body>
                <h1>Sanitary Conditions of Worshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="//fs[f[@select='Yes']]"/>
                </table>
                <h2>Sources of Offensive Odors</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Water Closet</th>
                        <th>Other Sources</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="//fs[descendant::f[not(@select='yes')]]"/>
                    
                </table>
                <h2>Standing or Sitting at Work</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Sit</th>
                        <th>Stand</th>
                    </tr>
                    <xsl:apply-templates select="//fs[descendant::f[@select='yes']]"/>
                </table>
            </body>
        </html>
    </xsl:template>
    <xsl:template match="fs[f[@select='Yes']]">
        <tr>
            <td>
                <xsl:value-of select="count(preceding::fs[f[@select='Yes']])+1"/>
            </td>
            <td>
                <xsl:apply-templates select="f[@name='question']/normalize-space()"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Yes']/@n"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Yes_but_fined']/@n"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='No']/@n"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Blank']/@n"/>
            </td>
            <td> <xsl:value-of select="sum(.//fs//descendant::f[@n])"/></td>
            
        </tr>
 </xsl:template>
    <xsl:template match= "fs[descendant::f[not(@select='yes')]]">
    <tr><td><xsl:apply-templates select="f[@name='question']/normalize-space()"/></td>
        <td><xsl:value-of select="f[@select='water_closet']/@n"/></td>
        <td><xsl:value-of select="f[@select='other_causes']/@n"/></td>
        <td> <xsl:value-of select="sum(.//fs//descendant::f[@n])"/></td>
    </tr>
    </xsl:template>
    <xsl:template match="fs[descendant::f[@select='yes']]">
        <tr><td><xsl:apply-templates select="f[@name='question']/normalize-space()"/></td>
            <td><xsl:value-of select="f[@select='Sit']/@n"/></td>
            <td><xsl:value-of select="f[@select='Stand']/@n"/></td>
            <td><xsl:value-of select="f[@select='Optional']/@n"/></td>    
      </tr>  
    </xsl:template>
</xsl:stylesheet>`

**Comment by setriplette** [comment]
I fixed my capitalization of Yes and it made it different, but not very much better.
`<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xpath-default-namespace="http://www.tei-c.org/ns/1.0"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    exclude-result-prefixes="xs math"
    xmlns="http://www.w3.org/1999/xhtml"
    version="3.0">
    
    <xsl:output method="xhtml" encoding="utf-8" doctype-system="about:legacy-compat"
        omit-xml-declaration="yes"/>
    <xsl:template match="/">
        
        <html>
            <head>
                <title>Sanitary Conditions NYC</title>
            </head>
            <body>
                <h1>Sanitary Conditions of Worshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="//fs[f[@select='Yes']]"/>
                </table>
                
                <h2>Sources of Offensive Odors</h2>
                
                <table>
                    <tr>
                        
                        <th>Number</th>
                        <th>Question</th>
                        <th>Water Closet</th>
                        <th>Other Sources</th>
                        <th>Total Responses</th>
                        
                    </tr>
                    <xsl:apply-templates select="//fs[descendant::f[not(@select='Yes')]]"/>
                    
                </table>
                
                
                <h2>Standing or Sitting at Work</h2>
                
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Sit</th>
                        <th>Stand</th>
                    </tr>
                    <xsl:apply-templates select="//fs[descendant::f[@select='Yes']]"/>
                </table>
            </body>
        </html>
    </xsl:template>
    
    <xsl:template match="fs[f[@select='Yes']]">
        
        <tr>
            <td>
                <xsl:value-of select="count(preceding::fs[f[@select='Yes']])+1"/>
            </td>
            <td>
                <xsl:apply-templates select="f[@name='question']/normalize-space()"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Yes']/@n"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Yes_but_fined']/@n"/>
            </td>
    
            
            <td>
                <xsl:value-of select="f[@select='No']/@n"/>
            </td>
            <td>
                <xsl:value-of select="f[@select='Blank']/@n"/>
            </td>
            <td> <xsl:value-of select="sum(.//fs//descendant::f[@n])"/></td>
            
        </tr>
 </xsl:template>
    <xsl:template match= "fs[descendant::f[not(@select='Yes')]]">
    <tr><td><xsl:apply-templates select="f[@name='question']/normalize-space()"/></td>
  
        <td><xsl:value-of select="f[@select='water_closet']/@n"/></td>
        <td><xsl:value-of select="f[@select='other_causes']/@n"/></td>
        <td> <xsl:value-of select="sum(.//fs//descendant::f[@n])"/></td>
    </tr>
    </xsl:template>
    <xsl:template match="fs[descendant::f[@select='Yes']]">
        <tr><td><xsl:apply-templates select="f[@name='question']/normalize-space()"/></td>
            <td><xsl:value-of select="f[@select='Sit']/@n"/></td>
            <td><xsl:value-of select="f[@select='Stand']/@n"/></td>
            <td><xsl:value-of select="f[@select='Optional']/@n"/></td>
          
      </tr>  
    </xsl:template>
</xsl:stylesheet>`

**Comment by RJP43** [comment]
@setriplette I went in and removed the bottom portion of your code to give others a chance to complete that portion of the assignment on their own and because your error is inside of your document node match.

**Comment by ebeshero** [comment]
@setriplette I was just typing that you need to notice that the values of the @select attributes are capitalized, but you changed it... So what is the error now?

**Comment by setriplette** [comment]
I'm not matching on the right questions in the bottom two tables--can't figure out how to match the sub-questions. 

Also about a hundred other things that I can't find. I deleted my code just in case.

**Comment by ebeshero** [comment]
@setriplette Notice that the `<f>` element isn't a descendant of fs. It's a child. That oughta help in that template match on the document node.

**Comment by setriplette** [comment]
What is the difference between child and descendant?

**Comment by RJP43** [comment]
@setriplette each of your apply-templates need to specify something different (in fact we will show you in a later assignment how to use something called modal xslt which will allow you to select the same parts of the XML) , but for this assignment that is not necessary since each table wants to process only fs elements that have f elements with certain responses. See how all of your tables are calling on `<xsl:apply-templates select="//fs[descendant::f[@select='Yes']]"/>` well thats not actually what you want for the tables about offensive odors or sitting/standing.

**Comment by ebeshero** [comment]
Child is one path step down. Just a child. Descendant should catch both child and grandchildren. It's probably not a big deal, actually.

**Comment by setriplette** [comment]
I replaced all my descendants with child and still have same problems.

**Comment by ebeshero** [comment]
@RJP43 Actually, no: One of her tables is  using the not() function: <xsl:apply-templates select="//fs[descendant::f[not(@select='yes')]]"/>

**Comment by RJP43** [comment]
@ebeshero and the other one isn't... soooo is it not still something important to mention.

**Comment by setriplette** [comment]
The subquestions seem to me to have the same xpath as the main questions

` <fs type="QandA">
              <f name="question">
                 <string>Are there offensive odors in the rooms occupied by employes[?]</string>
              </f>
              <f name="response" select="Yes" n="342"/>
              <f name="response" select="No" n="233"/>
              <f name="response" select="Blank" n="235"/>
           </fs>
            <fs type="QandA">
               <f name="question">
                  <string>[If there are offensive odors in the rooms], from what causes?</string>
               </f>
               <f name="response" select="water_closet" n="81"/>
               <f name="response" select="other_causes" n="73"/>
            </fs>`

How could I grab these two as separate question types when their code looks the same to me?

**Comment by setriplette** [comment]
They have `<fs type="QandA">`, `<f name="question"> and a string and appear to be siblings.

**Comment by ebeshero** [comment]
@setriplette Look at the `@select` attributes and their values.

**Comment by RJP43** [comment]
well they differ because the sibling f elements have different @select values

**Comment by ebeshero** [comment]
Yeah @RJP43 she didn't change the last table setup, but she doesn't need three tables anyway.

**Comment by setriplette** [comment]
So I need a separate table for each @select value?

**Comment by setriplette** [comment]
I could figure out the xpath to grab each @select value, like so

`//fs[child::f[@select="Yes"]]`

But not how to put the yes, no, and blank in one group and the words in another. Sigh.

**Comment by ebeshero** [comment]
@setriplette @RJP43 I just checked the assignment-- Yeah, you do want three tables, but here's why:

1. Table 1 is for the 12 questions that have yes or no or blank answers.
2. Table 2 is for the Water Closet question
3. Table three is for Sitting or Standing at work.

The reason for the three tables is because they don't have the same kinds of answers. You can tell this by looking at the `@select` attributes on these questions. Previously when I walked in cold from my office I was telling you these just were NOT "yes" answers. But that's what's leading to your ambiguous rule matches--sorry! On those two other tables, you want to be calling on ONLY the `<fs>` elements that contain child `<f>` elements whose `@select` is a specific value that makes them distinctive.

**Comment by ebeshero** [comment]
So for example:
````
<fs type="QandA">
<f name="question">
<string>
[If there are offensive odors in the rooms], from what causes?
</string>
</f>
<f name="response" select="water_closet" n="81"/>
<f name="response" select="other_causes" n="73"/>
</fs>
````

**Comment by RJP43** [comment]
From my comment above to @etj27: 

Your apply-templates for each table will need to grab only the information that pertains to each of those unique tables.  In the table where you are giving the data on yes and no questions you want to process the contents of the <fs> elements that have descendant <f> elements that specifically have the kinds of responses present in that table (so either Yes, No, or Yes_but_fined for the first table) So each table will have their own <xsl:apply-templates> selection that is specific to the kind of responses be represented by that table. In our class-example (which I added to significantly after class) notice I marked each template match (after the first document node match) to indicate what <xsl:apply-templates> selection from the document node it matches. Each table will get one <xsl:apply-templates> selection in the source node template match and a corresponding template match where you give the contents of the table row and each table cell. The <xsl:apply-templates> selection in the source document is where you say HEY this table is going to have this specific info processed in it and the corresponding xsl:template @match gives the directions of how that info is processed.

**Comment by RJP43** [comment]
So your first apply-templates selection in the source node match was right `<xsl:apply-templates select="//fs[descendant::f[@select='yes']]"/>` then in the next table you don't want the `<fs>` elements that have descendant `f[@select='yes']` instead you want the `<fs>` elements that have `<f>` elements with `@select` attributes that match offensive odor questions and then the same thing in the third table for the sit/stand questions.

**Comment by RJP43** [comment]
and as it seems you caught earlier it is very important that you mimic exactly the spelling of the attribute values from the XML  ... so `yes` won't work because in the XML the `@select` is equal to `Yes`

**Comment by setriplette** [comment]
The yes I could do because it is only the one thing. How do I write an or into these statements? Is it a comma, a pipe bar, something else?

**Comment by ebeshero** [comment]
@setriplette When the value of `@select` on the f element is "water_closet", how would you land on the fs that contains such a thing?  I think you're making this too complicated! You need no bars, pipes, or anything. Just notice that these f elements have a very special value that the others don't have.

**Comment by RJP43** [comment]
@setriplette you don't need bars, pipes, or anything  because in the XML if the `<fs>` has `<f>` elements with `select='Yes'` then its sibling `<f>` elements all correspond (No, Blank, Yes_but_fined) with the only logical responses to the question asked. Its not about grabbing all of them, but instead grabbing one of them so you can process each of the  different `<f>` elements in your template match for that table. Your apply-templates for each table in the document node match are grabbing the specific `<fs>` elements so then you can process the content (specific `<f>` elements) later in a template match for each table.

**Comment by setriplette** [comment]
I thought about trying a matches expression with the underscore, but now I'm realizing that the Yes_but_fined has it as well so that won't find it. I'm saying uncle and maybe we can see the answer in class tomorrow?

**Comment by RJP43** [comment]
@setriplette I think you can get this! Like I said above `<xsl:apply-templates select="//fs[descendant::f[@select='Yes']]"/>` was right for your first apply-templates for your first table. How could you change it slightly so you get fs elements that have descendant f elements where the @select is equal to other question responses?

**Comment by ebeshero** [comment]
@setriplette You don't need a function. You just need to modify your XPath expression--and it's so simple that I have *already shown you the answer* in this thread. I'll repeat it:
"When the value of @select on the f element is "water_closet", how would you land on the fs that contains such a thing? ...
Literally, just replace 'Yes' with the new value."

**Comment by ebeshero** [comment]
Try it--just try what I just told you right now, and see if it works, okay?

**Comment by RJP43** [comment]
@etj27 I created a screenshot from the in-class example that hopefully helps illustrate the one-to-one versus one-to-many setup of the document. @msb81 this may help you as well. 
So for each specific selection on the apply-templates for each table in the document node match you create a new template that matches on that selection. 
![screenshot_1](https://cloud.githubusercontent.com/assets/8645521/19705255/7feb9dc2-9ad9-11e6-9760-95454bc218c4.jpg)

**Comment by laurenmcguigan** [comment]
okay  I am lost on how to do the yes/no part. I have gotten this far but it is outputting what I want. For the yes, yes but fined, no, blank columns I am getting ones in random places.

````
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xpath-default-namespace="http://www.tei-c.org/ns/1.0"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    exclude-result-prefixes="xs math"
    xmlns="http://www.w3.org/1999/xhtml"
    version="3.0">
    
    <xsl:output method="xhtml" encoding="utf-8" doctype-system="about:legacy-compat"
        omit-xml-declaration="yes"/>
    
    <xsl:template match="/">
        <html>
            <head><title>Sanitary Conditions of Workshops and Factories in New York City</title></head>
            
            <body>
                <h1>Sanitary Conditions of Workshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="descendant::f[@name='question']"/>
                </table>
            </body>
        </html>
    </xsl:template>
<xsl:template match="f[@name='question']">
  <tr>
      <td><xsl:apply-templates select="descendant::string"/></td>
  </tr>
<tr>    
  <td>  <xsl:apply-templates select="following-sibling::f[@name='response']"/></td>
</tr>
</xsl:template>
    
    <xsl:template match="f[@name='response']">

           <td><xsl:apply-templates select="count(@select='Yes')"/></td>
       
    </xsl:template>
    
</xsl:stylesheet>
````

**Comment by laurenmcguigan** [comment]
I think this might be closer but I'm not sure...

````
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xpath-default-namespace="http://www.tei-c.org/ns/1.0"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    exclude-result-prefixes="xs math"
    xmlns="http://www.w3.org/1999/xhtml"
    version="3.0">
    
    <xsl:output method="xhtml" encoding="utf-8" doctype-system="about:legacy-compat"
        omit-xml-declaration="yes"/>
    
    <xsl:template match="/">
        <html>
            <head><title>Sanitary Conditions of Workshops and Factories in New York City</title></head>
            
            <body>
                <h1>Sanitary Conditions of Workshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="descendant::f[@name='question']"/>
                    <xsl:apply-templates select="descendant::f[@name='response']"/>
                </table>
            </body>
        </html>
    </xsl:template>
<xsl:template match="f[@name='question']">
  <tr>
      <td><xsl:apply-templates select="descendant::string"/></td>
  </tr>
</xsl:template>
    
    <xsl:template match="f[@name='response']">
<tr>
    <td><xsl:apply-templates select="count(@select='Yes')"/></td>
</tr>
   
       
    </xsl:template>
    
</xsl:stylesheet>
````

**Comment by RJP43** [comment]
@laurenmcguigan in my screenshot above from the in-class example each table only gets one apply-templates (in the template match on the document node) and then each of those apply-templates get matched on in their own template rule that contains **ALL** of the other table cells for that table. I think if you go back and read thru the rest of this issue you will see your fundamental structural issues.

**Comment by RJP43** [comment]
@laurenmcguigan your apply-template selections in each of the tables should not be grabbing the `<f>` elements but instead grabbing the `<fs>` elements **the have** specific descendent `<f>` elements defined by the `@select` attributes.

**Comment by laurenmcguigan** [comment]
@RJP43  SO does that mean this is right so far? or Do I need a new template rule for the yes?
````
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
    xmlns:xs="http://www.w3.org/2001/XMLSchema"
    xpath-default-namespace="http://www.tei-c.org/ns/1.0"
    xmlns:math="http://www.w3.org/2005/xpath-functions/math"
    exclude-result-prefixes="xs math"
    xmlns="http://www.w3.org/1999/xhtml"
    version="3.0">
    
    <xsl:output method="xhtml" encoding="utf-8" doctype-system="about:legacy-compat"
        omit-xml-declaration="yes"/>
    
    <xsl:template match="/">
        <html>
            <head><title>Sanitary Conditions of Workshops and Factories in New York City</title></head>
            
            <body>
                <h1>Sanitary Conditions of Workshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr>
                    <xsl:apply-templates select="descendant::f[@name='question']"/>
                </table>
            </body>
        </html>
    </xsl:template>
<xsl:template match="f[@name='question']">
  <tr>
      <td><xsl:apply-templates select="descendant::string"/></td>
      <td><xsl:value-of select="count(@select='Yes')"/></td>
  </tr>
</xsl:template>
</xsl:stylesheet>

````

**Comment by RJP43** [comment]
@laurenmcguigan Since you are just working with the first table, for now, then you only should have one apply-templates in that table structure that you have correctly created in the document node match. Then a single template that matches on what you selected in the apply-templates. So that part is right that you have one template matching the apply-templates from above. However you apply-template needs to grab the `<fs>` element not the specific `<f>` element because you will be grabbing the multiple `<f>` elements in the template match instead. Your apply-templates selection in the document node match needs to be more general so you can work with all of its specific contents.

**Comment by RJP43** [comment]
@laurenmcguigan see above where we told @setriplette that in order to process entire question results in the match the first apply-templates needs to just grab the `<fs>` elements that have descendent `<f>` elements that have a certain `@select` value that indicate the kind of questions you want to grab in that table. Since the `<f>` element that is associated with the questions do not give an attribute that indicate they are yes/no questions then you need to look at the sibling `<f>` elements and use an attribute found there to select certain `<fs>` elements that have those kind of responses.

**Comment by RJP43** [comment]
@laurenmcguigan it is important to realize that the first apply-templates selection (in the document node match) needs to allow the whole `<fs>` element to be processed so in the corresponding template match each `<td>` can contain an apply-templates selection that grabs all the different `<f>` elements.

**Comment by RJP43** [comment]
@laurenmcguigan Your apply-templates in each of the tables created in the match on the document node will have a predicate statement sitting inside a predicate statement because you want the `<fs>` elements **that have** descendant `<f>` elements **that have** a certain `@select` value indicating the kind of question (based on the response). Only one `@select` predicate is necessary because if the question has a `f[@select="No"]` then it also has a `f[@select="Yes"]` and a `f[@select="Yes_but_fined"]` and a `f[@select="Blank"]` since all of the responses for a question are in a single `<fs>` element and that is what you are stepping down to grab.

**Comment by laurenmcguigan** [comment]
@RJP43 Okay! I completely understand what you are saying now! I am just having one final problem. I can't get the values in their own row. They want to be in the same row as the questions.

**Comment by RJP43** [comment]
@laurenmcguigan you want it all to be in the same row (`<tr>`). Because it is dynamic each time a `<fs>` element comes up in the XML it will process all of the template match (which should include all of the apply-templates selections on the different `<f>` elements inside of `<td>` elements) for that question (`<fs>` element) and then go on to the next one.

**Comment by RJP43** [comment]
@laurenmcguigan Your write the `<tr>` element once and the`<td>` elements once for each element being processed (with apply-templates) and the computer does the work of looping through all of the `<fs>` elements as they appear in the XML.

**Comment by laurenmcguigan** [comment]
Okay, Im sorry I actually meant to say they are all in the same column like this

Is there free circulation of air through and about the building in which you work? One says where she is employed there are 100 women and small girls who work in a cellar without ventilation, and gas and electric lights burning all day.
468
249
93

Instead of being across the row

**Comment by laurenmcguigan** [comment]
@RJP43  NEVERMIND I FIGURED IT OUT SORRY IM DUMB

**Comment by RJP43** [comment]
@laurenmcguigan HUZZAH! and no way you are very smart! Thank you for posting when you got stuck.

**Comment by jonhoranic** [comment]
@RJP43 Ok so, I have the structure of most of this working, but I cannot seem to grab the correct values off the n attribute with anything I have tried thus far, all it does is display a 1 on the table for each cell down the columns.
This would be everything for table one:
`<h1>Sanitary Conditions of Worshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr><xsl:apply-templates select="descendant::fs[f[@select='Yes']]"/>
                </table>`
`<xsl:template match="fs[f[@select='Yes']]">
        <tr>
            <td><xsl:value-of select="count(preceding-sibling::fs)+1"/></td>
            <td><xsl:apply-templates select="descendant::string"/></td>
            <td><xsl:value-of select="count(f[@select='Yes'])"/></td>
            <td></td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </xsl:template>`

**Comment by jonhoranic** [comment]
@RJP43 I have yet to fill out the rest until I can get around displaying the proper numbers, so am I missing a certain function? Or is this a matching problem that I just need to refine?

**Comment by RJP43** [comment]
Evening (well morning actually) @jonhoranic 
 Your template matches look good as well as you initial apply templates as long as that is appearing in the template match on the document node.
In your `<td>` that you want the number of Yes responses to show up you are complicating it with the count function. You simply want to walk the tree down to the `@n` for each response. So instead of `<td><xsl:value-of select="count(f[@select='Yes'])"/>` you want something like this: `<td><xsl:apply-templates select="element[@select='desired response']/@n"/>

**Comment by RJP43** [comment]
@jonhoranic is this part of your code: 
````
<h1>Sanitary Conditions of Worshops and Factories in New York City</h1> <h2>Table 1: Yes or No Questions</h2> <table> <tr> <th>Number</th> <th>Question</th> <th>Yes</th> <th>Yes, but fined</th> <th>No</th> <th>Blank</th> <th>Total Responses</th> </tr><xsl:apply-templates select="descendant::fs[f[@select='Yes']]"/> </table>
````
inside of a template that matches on the document node?

**Comment by jonhoranic** [comment]
@RJP43 Let me try to make that more legible.

`<xsl:template match="/">
        <html>
            <head>
                <title>NYC "Working Area Sanitary Conditions" Tables</title>
                <style>
                    table, th, td {
                    border: 1px solid black;
                    }
                </style>
            </head>
            <body>
                <h1>Sanitary Conditions of Worshops and Factories in New York City</h1>
                <h2>Table 1: Yes or No Questions</h2>
                <table>
                    <tr>
                        <th>Number</th>
                        <th>Question</th>
                        <th>Yes</th>
                        <th>Yes, but fined</th>
                        <th>No</th>
                        <th>Blank</th>
                        <th>Total Responses</th>
                    </tr><xsl:apply-templates select="descendant::fs[f[@select='Yes']]"/>
                </table>
                <h2>Table 2: Sources of Offensive Odors</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Water Closet</th>
                        <th>Other Sources</th>
                        <th>Total Responses</th>
                    </tr><xsl:apply-templates select="descendant::fs[descendant::f[@select='water_closet']]"/>
                </table>
                <h2>Table 3: Standing or Sitting at Work</h2>
                <table>
                    <tr>
                        <th>Question</th>
                        <th>Sit</th>
                        <th>Stand</th>
                        <th>Optional</th>
                        <th>Blank</th>
                        <th>Total Responces</th>
                    </tr><xsl:apply-templates select="descendant::fs[descendant::f[@select='Sit']]"/>
                </table>
            </body>
        </html>
    </xsl:template>
    
    <xsl:template match="fs[f[@select='Yes']]">
        <tr>
            <td><xsl:value-of select="count(preceding-sibling::fs)+1"/></td>
            <td><xsl:apply-templates select="descendant::string"/></td>
            <td><xsl:apply-templates select="@n"/></td>
            <td></td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </xsl:template>
    
    <xsl:template match="fs[f[@select='water_closet']]">
        <tr>
            <td><xsl:apply-templates select="descendant::string"/></td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </xsl:template>
    
    <xsl:template match="fs[f[@select='Sit']]">
        <tr>
            <td><xsl:apply-templates select="descendant::string"/></td>
            <td></td>
            <td></td>
            <td></td>
            <td></td>
            <td></td>
        </tr>
    </xsl:template>
    
</xsl:stylesheet>`

**Comment by jonhoranic** [comment]
@RJP43 Something tells me that is not supposed to happen when posting code on the GitHub. I have no idea what is causing that!

**Comment by RJP43** [comment]
I fixed it... just have to drop four tick marks in front and at the end when you have long chunks of code.

**Comment by RJP43** [comment]
@jonhoranic Okay soooo yes this is all looking good so far until you get to the `<td>` element that you are trying to grab the `@n` values for each response. `<td><xsl:apply-templates select="@n"/></td>` won't work because you haven't said what `<f>` element in the `<fs>` your template is matching on that you want the `@n` value of. So if this `<td>` is supposed to be the `@n` value of the `<f>` elements that are the yes responses you need to step to from the `<fs>` element to the `<f>` element **that have** `@select='Yes'` and then take a step down to the `@n`

**Comment by RJP43** [comment]
@jonhoranic you see why that is? If your match is grabbing the `<fs>` elements all of your table cells need to step down into the different parts of the `<fs>` where you want them.

**Comment by jonhoranic** [comment]
@RJP43 I think so? It would look something like this then right?
`<td><xsl:apply-templates select="fs[f[@select='Yes']]/@n"/></td>`

**Comment by RJP43** [comment]
@jonhoranic well you don't want to call on the fs element because that is where you are coming from already according to your template match... you literally want what I have above. You step from your template match to the elements and its attributes you want to grab in each `<td>`.

**Comment by RJP43** [comment]
So actually all of your `<td>` elements need revised so you walk the tree to the element that has the content you want in that table cell.

**Comment by jonhoranic** [comment]
@RJP43 Ok I see now! Though I just noticed a new issue, my Number column is off on the table's HTML view and I do not know why that would be? It counts 1, 2, 4, 5, 6, 7, 8, 9, 10, 11, 12,14

**Comment by RJP43** [comment]
@jonhoranic 
Now the first table cell is a little complicated, but you were already very close. For that cell you need to say count from where I am (self or dot-notation) the preceding-sibling `<fs>` **of the same kind of question** and add 1. What you had above was saying grab any preceding-sibling `<fs>` ... make sense?

**Comment by RJP43** [comment]
@jonhoranic So the same way in the template match that you defined the specific kind of question by the `f[@select]` you will want to do that here to define the kind of preceding-sibling `<fs>` element you want to count.

**Comment by jonhoranic** [comment]
@RJP43  All good on that, and now finally my No column is not displaying anything even though it is typed like the other lines

**Comment by RJP43** [comment]
@jonhoranic let's see what you have for that `<td>` ???

**Comment by jonhoranic** [comment]
@RJP43 Nevermind! It was a tad broke and I did not notice, all good now!
